import the libaries

Read the dataset

Perform an initial audit (# Shape

5. Save a Data Quality Snapshot
Create a markdown cell in the notebook and write observations as you go.
6. import the libaries
Read the dataset
Perform an initial audit
Then check missing values

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [52]:
df = pd.read_csv("Customer_support_data.csv")

In [3]:
# Shape

df.shape

(85907, 20)

In [4]:
# First five rows

df.head()

,Unique id,channel_name,category,Sub-category,Customer Remarks,Order_id,order_date_time,Issue_reported at,issue_responded,Survey_response_Date,Customer_City,Product_category,Item_price,connected_handling_time,Agent_name,Supervisor,Manager,Tenure Bucket,Agent Shift,CSAT Score
0,7e9ae164-6a8b-4521-a2d4-58f7c9fff13f,Outcall,Product Queries,Life Insurance,NaN,c27c9bb4-fa36-4140-9f1f-21009254ffdb,NaN,01-08-2023 11:13,01-08-2023 11:47,01-Aug-23,NaN,NaN,NaN,NaN,Richard Buchanan,Mason Gupta,Jennifer Nguyen,On Job Training,Morning,5
1,b07ec1b0-f376-43b6-86df-ec03da3b2e16,Outcall,Product Queries,Product Specific Information,NaN,d406b0c7-ce17-4654-b9de-f08d421254bd,NaN,01-08-2023 12:52,01-08-2023 12:54,01-Aug-23,NaN,NaN,NaN,NaN,Vicki Collins,Dylan Kim,Michael Lee,>90,Morning,5
2,200814dd-27c7-4149-ba2b-bd3af3092880,Inbound,Order Related,Installation/demo,NaN,c273368d-b961-44cb-beaf-62d6fd6c00d5,NaN,01-08-2023 20:16,01-08-2023 20:38,01-Aug-23,NaN,NaN,NaN,NaN,Duane Norman,Jackson Park,William Kim,On Job Training,Evening,5
3,eb0d3e53-c1ca-42d3-8486-e42c8d622135,Inbound,Returns,Reverse Pickup Enquiry,NaN,5aed0059-55a4-4ec6-bb54-97942092020a,NaN,01-08-2023 20:56,01-08-2023 21:16,01-Aug-23,NaN,NaN,NaN,NaN,Patrick Flores,Olivia Wang,John Smith,>90,Evening,5
4,ba903143-1e54-406c-b969-46c52f92e5df,Inbound,Cancellation,Not Needed,NaN,e8bed5a9-6933-4aff-9dc6-ccefd7dcde59,NaN,01-08-2023 10:30,01-08-2023 10:32,01-Aug-23,NaN,NaN,NaN,NaN,Christopher Sanchez,Austin Johnson,Michael Lee,0-30,Morning,5


In [5]:
# Column names
print(df.columns)

Index(['Unique id', 'channel_name', 'category', 'Sub-category',
       'Customer Remarks', 'Order_id', 'order_date_time', 'Issue_reported at',
       'issue_responded', 'Survey_response_Date', 'Customer_City',
       'Product_category', 'Item_price', 'connected_handling_time',
       'Agent_name', 'Supervisor', 'Manager', 'Tenure Bucket', 'Agent Shift',
       'CSAT Score'],
      dtype='object')


In [6]:
# Data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85907 entries, 0 to 85906
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unique id                85907 non-null  object 
 1   channel_name             85907 non-null  object 
 2   category                 85907 non-null  object 
 3   Sub-category             85907 non-null  object 
 4   Customer Remarks         28742 non-null  object 
 5   Order_id                 67675 non-null  object 
 6   order_date_time          17214 non-null  object 
 7   Issue_reported at        85907 non-null  object 
 8   issue_responded          85907 non-null  object 
 9   Survey_response_Date     85907 non-null  object 
 10  Customer_City            17079 non-null  object 
 11  Product_category         17196 non-null  object 
 12  Item_price               17206 non-null  float64
 13  connected_handling_time  242 non-null    float64
 14  Agent_name            

In [7]:
# Summary statistics
df.describe(include="all")

,Unique id,channel_name,category,Sub-category,Customer Remarks,Order_id,order_date_time,Issue_reported at,issue_responded,Survey_response_Date,Customer_City,Product_category,Item_price,connected_handling_time,Agent_name,Supervisor,Manager,Tenure Bucket,Agent Shift,CSAT Score
count,85907,85907,85907,85907,28742,67675,17214,85907,85907,85907,17079,17196,17206.000000,242.000000,85907,85907,85907,85907,85907,85907.000000
unique,85907,3,12,57,18231,67675,13766,30923,30262,31,1782,9,NaN,NaN,1371,40,6,5,5,NaN
top,7e9ae164-6a8b-4521-a2d4-58f7c9fff13f,Inbound,Returns,Reverse Pickup Enquiry,Good,c27c9bb4-fa36-4140-9f1f-21009254ffdb,09-08-2023 11:55,15-08-2023 10:59,28-08-2023 00:00,28-Aug-23,HYDERABAD,Electronics,NaN,NaN,Wendy Taylor,Carter Park,John Smith,>90,Morning,NaN
freq,1,68142,44097,22389,1390,1,7,13,3378,3452,722,4706,NaN,NaN,429,4273,25261,30660,41426,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5660.774846,462.400826,NaN,NaN,NaN,NaN,NaN,4.242157
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12825.728411,246.295037,NaN,NaN,NaN,NaN,NaN,1.378903
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,NaN,NaN,NaN,NaN,NaN,1.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,392.000000,293.000000,NaN,NaN,NaN,NaN,NaN,4.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,979.000000,427.000000,NaN,NaN,NaN,NaN,NaN,5.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2699.750000,592.250000,NaN,NaN,NaN,NaN,NaN,5.000000


In [54]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
string_columns = list(df.dtypes[df.dtypes == 'object'].index)
for col in string_columns:
    df[col] = df[col].str.lower().str.replace(' ', '_')

In [ ]:
#Missing value

missing = df.isna().sum().to_frame("Missing Values")

missing["Percentage"] = (missing["Missing Values"] / len(df) * 100).round(2)

missing.sort_values("Missing Values", ascending=False)

,Missing Values,Percentage
connected_handling_time,85665,99.72
Customer_City,68828,80.12
Product_category,68711,79.98
Item_price,68701,79.97
order_date_time,68693,79.96
Customer Remarks,57165,66.54
Order_id,18232,21.22
Agent Shift,0,0.00
Tenure Bucket,0,0.00
Manager,0,0.00


In [78]:
#Duplicate chaeck
duplicates = df["unique_id"].duplicated().sum()

print(duplicates)

0


In [31]:
df.columns

Index(['unique_id', 'channel_name', 'category', 'sub-category',
       'customer_remarks', 'order_id', 'order_date_time', 'issue_reported_at',
       'issue_responded', 'survey_response_date', 'customer_city',
       'product_category', 'item_price', 'connected_handling_time',
       'agent_name', 'supervisor', 'manager', 'tenure_bucket', 'agent_shift',
       'csat_score'],
      dtype='object')

In [32]:
# Check how many unique timestamp patterns exist
df["issue_reported_at"].sample(20, random_state=42)

67871    26-08-2023_18:39
40187    14-08-2023_00:32
60075    22-08-2023_20:48
69560    26-08-2023_23:47
2605     01-08-2023_18:37
73327    27-08-2023_14:28
4382     02-08-2023_19:54
10405    06-08-2023_12:24
24494    10-08-2023_21:52
5473     03-08-2023_02:04
16935    07-08-2023_19:07
23635    10-08-2023_19:07
81456    30-08-2023_20:11
10045    06-08-2023_13:32
77190    28-08-2023_11:09
4154     01-08-2023_19:02
65466    25-08-2023_15:48
40932    16-08-2023_10:24
30817    12-08-2023_06:52
6578     03-08-2023_09:02
Name: issue_reported_at, dtype: object

In [35]:
df["issue_reported_at"].str.contains("/", na=False).sum()

0

In [37]:
df["issue_reported_at"].str.contains("-", na=False).sum()

85907

In [55]:
df2 = df.copy()

In [56]:
df["issue_reported_at"] = (
    df["issue_reported_at"]
      .str.replace("_", " ", regex=False)
)

df["issue_responded"] = (
    df["issue_responded"]
      .str.replace("_", " ", regex=False)
)

In [57]:
df[['issue_reported_at','issue_responded']].head(10)

,issue_reported_at,issue_responded
0,01-08-2023 11:13,01-08-2023 11:47
1,01-08-2023 12:52,01-08-2023 12:54
2,01-08-2023 20:16,01-08-2023 20:38
3,01-08-2023 20:56,01-08-2023 21:16
4,01-08-2023 10:30,01-08-2023 10:32
5,01-08-2023 15:13,01-08-2023 18:39
6,01-08-2023 15:31,01-08-2023 23:52
7,01-08-2023 16:17,01-08-2023 16:23
8,01-08-2023 21:03,01-08-2023 21:07
9,01-08-2023 23:31,01-08-2023 23:36


In [58]:
df["issue_reported_at"] = pd.to_datetime(
    df["issue_reported_at"],
    format="%d-%m-%Y %H:%M",
    errors="coerce"
)

df["issue_responded"] = pd.to_datetime(
    df["issue_responded"],
    format="%d-%m-%Y %H:%M",
    errors="coerce"
)

In [45]:
df["issue_reported_at"].isna().sum()

0

In [46]:
df["issue_responded"].isna().sum()

0

In [59]:
df["response_time_minutes"] = (
    df["issue_responded"] -
    df["issue_reported_at"]
).dt.total_seconds() / 60

In [48]:
df["response_time_minutes"].describe()

count    85907.000000
mean       136.886610
std        592.528551
min      -1437.000000
25%          2.000000
50%          5.000000
75%         35.000000
max       5758.000000
Name: response_time_minutes, dtype: float64

In [60]:
# Number of negative response times
negative_count = (df["response_time_minutes"] < 0).sum()

# Percentage of dataset
negative_percentage = (
    negative_count / len(df)
) * 100

print(negative_count)
print(round(negative_percentage, 2))

3128
3.64


In [ ]:
negative_rows = df.loc[
    df["response_time_minutes"] < 0,
    [
        "issue_reported_at",
        "issue_responded",
        "response_time_minutes"
    ]
]

negative_rows.head(20)

,issue_reported_at,issue_responded,response_time_minutes
74172,2023-08-28 16:07:00,2023-08-28,-967.0
74173,2023-08-28 16:56:00,2023-08-28,-1016.0
74174,2023-08-28 20:14:00,2023-08-28,-1214.0
74175,2023-08-28 14:37:00,2023-08-28,-877.0
74176,2023-08-28 23:26:00,2023-08-28,-1406.0
74177,2023-08-28 19:50:00,2023-08-28,-1190.0
74178,2023-08-28 07:42:00,2023-08-28,-462.0
74179,2023-08-28 20:00:00,2023-08-28,-1200.0
74180,2023-08-28 16:55:00,2023-08-28,-1015.0
74181,2023-08-28 20:31:00,2023-08-28,-1231.0


In [62]:
negative_rows

,issue_reported_at,issue_responded,response_time_minutes
74172,2023-08-28 16:07:00,2023-08-28,-967.0
74173,2023-08-28 16:56:00,2023-08-28,-1016.0
74174,2023-08-28 20:14:00,2023-08-28,-1214.0
74175,2023-08-28 14:37:00,2023-08-28,-877.0
74176,2023-08-28 23:26:00,2023-08-28,-1406.0
...,...,...,...
77566,2023-08-28 13:25:00,2023-08-28,-805.0
77567,2023-08-28 18:09:00,2023-08-28,-1089.0
77568,2023-08-28 22:20:00,2023-08-28,-1340.0
77569,2023-08-28 14:38:00,2023-08-28,-878.0


In [75]:
df2[74170:77573][['issue_reported_at','issue_responded']]

,issue_reported_at,issue_responded
74170,27-08-2023_14:16,27-08-2023_14:19
74171,27-08-2023_14:25,27-08-2023_14:27
74172,28-08-2023_16:07,28-08-2023_00:00
74173,28-08-2023_16:56,28-08-2023_00:00
74174,28-08-2023_20:14,28-08-2023_00:00
...,...,...
77568,28-08-2023_22:20,28-08-2023_00:00
77569,28-08-2023_14:38,28-08-2023_00:00
77570,28-08-2023_12:12,28-08-2023_00:00
77571,29-08-2023_15:06,29-08-2023_15:07


In [74]:
df2[['issue_reported_at','issue_responded']]

,issue_reported_at,issue_responded
0,01-08-2023_11:13,01-08-2023_11:47
1,01-08-2023_12:52,01-08-2023_12:54
2,01-08-2023_20:16,01-08-2023_20:38
3,01-08-2023_20:56,01-08-2023_21:16
4,01-08-2023_10:30,01-08-2023_10:32
...,...,...
85902,30-08-2023_23:20,31-08-2023_07:22
85903,31-08-2023_08:15,31-08-2023_08:17
85904,31-08-2023_18:57,31-08-2023_19:02
85905,31-08-2023_19:59,31-08-2023_20:00


In [76]:
df["response_time_status"] = np.where(
    df["response_time_minutes"] < 0,
    "Invalid",
    "Valid"
)

In [79]:
df.columns

Index(['unique_id', 'channel_name', 'category', 'sub-category',
       'customer_remarks', 'order_id', 'order_date_time', 'issue_reported_at',
       'issue_responded', 'survey_response_date', 'customer_city',
       'product_category', 'item_price', 'connected_handling_time',
       'agent_name', 'supervisor', 'manager', 'tenure_bucket', 'agent_shift',
       'csat_score', 'response_time_minutes', 'response_time_status'],
      dtype='object')

In [80]:
df["customer_city"] = df["customer_city"].fillna("Unknown")

df["product_category"] = df["product_category"].fillna("Unknown")

In [81]:
# Because only 242 out of 85,907 rows contain values (less than 0.3%), I'd remove it:
df.drop(columns=["connected_handling_time"], inplace=True)


In [82]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85907 entries, 0 to 85906
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   unique_id              85907 non-null  object        
 1   channel_name           85907 non-null  object        
 2   category               85907 non-null  object        
 3   sub-category           85907 non-null  object        
 4   customer_remarks       28742 non-null  object        
 5   order_id               67675 non-null  object        
 6   order_date_time        17214 non-null  object        
 7   issue_reported_at      85907 non-null  datetime64[ns]
 8   issue_responded        85907 non-null  datetime64[ns]
 9   survey_response_date   85907 non-null  object        
 10  customer_city          85907 non-null  object        
 11  product_category       85907 non-null  object        
 12  item_price             17206 non-null  float64       
 13  a

In [83]:

df.isnull().sum()

unique_id                    0
channel_name                 0
category                     0
sub-category                 0
customer_remarks         57165
order_id                 18232
order_date_time          68693
issue_reported_at            0
issue_responded              0
survey_response_date         0
customer_city                0
product_category             0
item_price               68701
agent_name                   0
supervisor                   0
manager                      0
tenure_bucket                0
agent_shift                  0
csat_score                   0
response_time_minutes        0
response_time_status         0
dtype: int64

In [84]:
df.describe()

,issue_reported_at,issue_responded,item_price,csat_score,response_time_minutes
count,85907,85907,17206.000000,85907.000000,85907.000000
mean,2023-08-16 22:29:05.784394752,2023-08-17 00:45:58.980990976,5660.774846,4.242157,136.886610
min,2023-07-28 20:42:00,2023-08-01 00:00:00,0.000000,1.000000,-1437.000000
25%,2023-08-09 12:48:00,2023-08-09 15:02:30,392.000000,4.000000,2.000000
50%,2023-08-16 18:22:00,2023-08-16 20:57:00,979.000000,5.000000,5.000000
75%,2023-08-24 17:33:30,2023-08-24 19:27:00,2699.750000,5.000000,35.000000
max,2023-08-31 23:58:00,2023-08-31 23:59:00,164999.000000,5.000000,5758.000000
std,NaN,NaN,12825.728411,1.378903,592.528551


In [86]:
# How many records have an Order ID?
df["order_id"].notna().sum()

67675

In [87]:
# Of the records with an Order ID, how many have an Order Date?
df.loc[df["order_id"].notna(), "order_date_time"].notna().sum()

17214

In [88]:
# Of the records with an Order ID, how many have an Item Price?
df.loc[df["order_id"].notna(), "item_price"].notna().sum()

17206

67675
17214
17206

In [91]:
df.head(6).T

,0,1,2,3,4,5
unique_id,7e9ae164-6a8b-4521-a2d4-58f7c9fff13f,b07ec1b0-f376-43b6-86df-ec03da3b2e16,200814dd-27c7-4149-ba2b-bd3af3092880,eb0d3e53-c1ca-42d3-8486-e42c8d622135,ba903143-1e54-406c-b969-46c52f92e5df,1cfde5b9-6112-44fc-8f3b-892196137a62
channel_name,outcall,outcall,inbound,inbound,inbound,email
category,product_queries,product_queries,order_related,returns,cancellation,returns
sub-category,life_insurance,product_specific_information,installation/demo,reverse_pickup_enquiry,not_needed,fraudulent_user
customer_remarks,NaN,NaN,NaN,NaN,NaN,NaN
order_id,c27c9bb4-fa36-4140-9f1f-21009254ffdb,d406b0c7-ce17-4654-b9de-f08d421254bd,c273368d-b961-44cb-beaf-62d6fd6c00d5,5aed0059-55a4-4ec6-bb54-97942092020a,e8bed5a9-6933-4aff-9dc6-ccefd7dcde59,a2938961-2833-45f1-83d6-678d9555c603
order_date_time,NaN,NaN,NaN,NaN,NaN,NaN
issue_reported_at,2023-08-01 11:13:00,2023-08-01 12:52:00,2023-08-01 20:16:00,2023-08-01 20:56:00,2023-08-01 10:30:00,2023-08-01 15:13:00
issue_responded,2023-08-01 11:47:00,2023-08-01 12:54:00,2023-08-01 20:38:00,2023-08-01 21:16:00,2023-08-01 10:32:00,2023-08-01 18:39:00
survey_response_date,01-aug-23,01-aug-23,01-aug-23,01-aug-23,01-aug-23,01-aug-23


In [93]:
df.to_csv("customer_support_clean.csv",index=False)

| Metric                   |                    Result |
| ------------------------ | ------------------------: |
| Total Rows               |                    85,907 |
| Total Columns            |                        21 |
| Duplicate Ticket IDs     |                         0 |
| Missing Customer Remarks |            57,165 (66.5%) |
| Missing Order IDs        |            18,232 (21.2%) |
| Missing Order Date       |            68,693 (80.0%) |
| Missing Item Price       |            68,701 (80.0%) |
| Removed Columns          | `connected_handling_time` |
| Invalid Response Times   |           3.64% (flagged) |
| Date Parsing             |              Standardized |
